packages/flutter/lib/src/semantics/semantics.dart

```diff
@@ -2776,6 +2782,95 @@ class SemanticsOwner extends ChangeNotifier {
   String toString() => describeIdentity(this);
 }
 
+/// A string that carries a list of [StringAttribute].
+@immutable
+class AttributedString {
+  /// Created Attributed string.
+  ///
+  /// the [TextRange] in the [attributes] must be in side the length of the
+  /// [string].
+  const AttributedString(
+    this.string, {
+    this.attributes,
+  });
+
+  /// The plain string.
+  final String string;
+
+  /// The attributes this string carries.
+  final List<StringAttribute>? attributes;
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType)
+      return false;
+    return other is AttributedString
+        && other.string == string
+        && other.attributes == attributes;
+  }
+
+  @override
+  int get hashCode => string.hashCode ^ attributes.hashCode;
+}
+
+/// The attribute type that is used in [StringAttribute].
+enum StringAttributeType {
+  /// The string should be spell out character by character.
+  spellOut,
+}
+
+/// The attribute that affects how assistive technologies treat the text.
+class StringAttribute {
+  /// Creates a text attribtue.
+  ///
+  /// The [range] specifies the portion in the text that carries this attribute
+  ///
+  /// Sets [StringAttributeType.spellOut] to true if you want the voiceover or talkback to spell out
+  /// the text character by character.
+  StringAttribute({
+    required TextRange range,
+    required this.type,
+    this.args,
+  }) : assert(range.isValid && !range.isCollapsed),
+       _range = range;
+
+  /// The text range to which this attribute should apply.
```

Comment:
What are these indexes representing? E.g. do the four byte emojis advance the index by one?

Reply:
emojis will advance index by 2~4, and I tested locally it looks like both ios and android expect so too. I think the current code is fine